README.md
# Cuca - a web application framework

Cuca is another web application framework written in Ruby. It's made to
build applications with focus on functionality - less design. You can
create and application without writing HTML. Once written Widgets
(a smart screen element) can be reused thorough your project with 
minimum effort allowing to build fast and secure web applications.

It implements the following concepts:
* A Widget is a screen element. Can be a full page or part of it.
   The Controller and the Layout are Widgets, too.
* A Controller deals with one request URI (get, post or both) and can set variables
   other widgets can make use of. It can also define a Layout and filters.
* A Layout wraps the output of a controller and finally return the
   built web page.
* A Generator (NOT "code generator") can be used within any Widget to help building the web content.
   Cuca comes with a Markaby and eruby Generator.
* A Session can used optionally to keep stateful data.

Cuca was written for internal use in 2008 and as per 2013 we keep using and improving it.

### It implements the following concepts:
* A Widget is a screen element. Can be a full page or part of it.
  The Controller and the Layout are Widgets, too.
* A Controller deals with one request URI (get, post or both) and can set variables
  other widgets can make use of. It can also define a Layout and filters.
* A Layout wraps the output of a controller and finally return the
  built web page.
* A Generator (NOT "code generator") can be used within any Widget to help building the web content.
 Cuca comes with a Markaby and eruby Generator.
* A Session can used optionally to keep stateful data.

## Installation & Getting started

 Download and install from the internet:

    gem install --remote cuca

## Create an application skeleton:

    cuca my_project		# this will create a new project
    cd my_project
    rackup

  Open http://localhost:9292/


## Read on

  Check the rdoc once installed



CHANGELOG
* 2021/02
   - Urlmap rewrite

* 2020/12 - Migrated to Rack, support until ruby 2.7

* 2014/12/31 - 0.11
   - Few fixes & cleanup
   - Ruby 2.1.0 support

* 2013/xx/xx - 0.10
   - Ruby 1.9 support
   - functional tests

* 2009 - 2013 ... big gap of changes done without publishing anything... ;)

* 2009/xx/xx - 0.07
   - 'stop' option :cancel_execution added. This can be raised in before filters
     and will avoid that the run/get/post... methods are called.
   - ARFormWidget will set attribites by the setter method which allows
     overwriting within the model
   - Minor Bugfixes
   - Added subcall? method to controller to easily detect subcalls


* 2009/03/04 - 0.06
   - Fixed sorting bug of filters with priority specified
   - Nicer pagination on listwidget for large lists
   - Allowing to tune default query definitions on listwidget 
   - Fixed bad rescue of NameError which resulted in misleading exception tracebacks
   - Catch exceptions on loading/parsing the controller code
   - Changed FormWidget and ARFormWidget, added FormElements module
   - 'stop' can switch off after filters with :no_after_filters option
   - Console will automatically load root support files
   - Delete method for session.page variables

* 2008/08/20 - 0.05
   - Widget definition can be done within the action controller file.
   - Fixed minor typos in application skeleton conf/config.rb
   - Fix to catch exceptions in conf/config and conf/environment
   - Fixes & minor changes in stdlib/form
   - Stdlib listwidget rewrite_hooks will pass named hash instead of plain array 
     as row data
   - fixed script/console to load support files

* 2008/06/17 - 0.04
   - Fixed error-message displaying traceback but not the actual exception
   - Display trace if enabled on 'stop :error' events
   - Option for autoloading support files (instead of bulk 'require')
   - Fixed bug where session didn't close (saved to disk) if 'stop' was raised (fastcgi)
   - 'url', 'base_path and 'path_tree' property added into urlmap class
   - Priority option on filters.
   - Mount option for external directories on application path
   - Cleanup app.rb, removed some global variables
   - get/post parameters are automaticaly visible within the session page container
   - Added some unit tests
   - Minor fixes in stdlib
   - Some fixes in dispatchers


* 2008/05/18 - 0.03
   - Fixed uncaught possible routing error exception
   - Splitted load configuration to conf/config.rb and conf/environment.rb
   - Allow setting the mime-type and http return status per controller action
   - Added http 404 and 500 customizable error messages
   - Option to switch off error tracebacks displayed in browser
   - 'base_url' property added into the urlmap class
   - Added 'layout' instance method within controller to overwrite 
     or disable default layout for current action
   - fixed bug on fastcgi where the 'stop' method could remove the layout 
     definition from controller class
   - Minor changes/fixes in application skeleton

* 2008/05/11 - 0.02
   - Fixed bug in erb/view generator ("can't find view path")
   - stdlib/listwidget:
         - Fixed table joins not displaying content (dblist)
         - Dynamic column sorting in both directions (ASC/DESC)
   - Fixed session bug when using fastcgi
   - minor documentation and stdlib improvements 

* 2008/05/08 - 0.01 - Initial Release


lib/cuca/app.rb
require 'logger'
require 'cuca/config'
require 'cuca/urlmap2'
require 'concurrent'

$cuca = Concurrent::ThreadLocalVar.new({})

def cuca_register_autoload(object, file)
 autoload(object, file)
end

module Cuca

# Sandbox is used internally run the action script defined by the controller
# In future this can be extended to implement some security features etc..
    class Sandbox

        def self.run(controller_class_name, mod, assigns, request_method, subcall)
            self.class.send(:include, mod)
            controller_class = mod::const_get(controller_class_name)
            controller = controller_class.send(:new, :assigns=>assigns)
            $cuca.value[:controller_object] = controller
        
            controller.run_before_filters
            case request_method
                when 'POST'
                    controller.send('_do'.intern, 'run', subcall)
                    controller.send('_do'.intern, 'post', subcall)
                when 'GET'
                    controller.send('_do'.intern, 'run', subcall)
                    controller.send('_do'.intern, 'get', subcall)
                when 'PUT'
                    controller.send('_do'.intern, 'put', subcall)
                when 'PATCH'
                    controller.send('_do'.intern, 'patch', subcall)
                when 'HEAD'
                    controller.send('_do'.intern, 'head', subcall)
                when 'DELETE'
                    controller.send('_do'.intern, 'delete', subcall)
                when 'OPTIONS'
                    controller.send('_do'.intern, 'options', subcall)        
            end
            controller.run_after_filters
            return [controller.http_status, controller.mime_type, controller.to_s, controller.http_header]
        end
    end # class Sandbox

    # == Cuca Application
    #
    # A Cuca::App object will be created directly by the dispatcher - which again is the
    # direct cgi or fastcgi script that get run by the webserver.
    # Normally you just create a Cuca::App object and run the cgicall method with optional with
    # a cgi object as paramenter.
    # Before doing that you must set $cuca_path to the root of your application directory structure.
    class App


        attr_reader :app_path, :log_path, :public_path, :url_scan, :cgi, :logger
        def logger=(log); @logger = log ; end
        @@app_config = Cuca::Config.new

        ## Application configuration
        public
        def self.configure
            yield @@app_config
        end

        def self.config
            @@app_config
        end

        # will do a 'require' on all .rb files in path
        def include_files(path)
            return unless File.exist?(path)
            pwd = Dir.pwd
            Dir.chdir(path)
            Dir['*.rb'].each do |f|
                require "#{path}/#{f}"
            end
            Dir.chdir(pwd)
        end

        # this will schedule all files for autoloading
        def autoload_file(node, naming_proc)
            path = node.value[:path]
            fn = path.scan(/.*\/(.*)\.rb/)[0][0]
            classname = naming_proc.call(fn)
            logger.debug "Scheduling Autoload Object #{classname.intern.inspect} ==> #{path}"

            cuca_register_autoload(classname.intern, clean_path(path))
        end

        # removes double slashes from a path-string
        def clean_path(directory)
            directory.gsub(/\/\//, '/')
        end

        def initialize(app_path: nil, log_path: nil, public_path: nil, additional_support_directory:nil)
            @app_path    = app_path || ($cuca_path + "/app")
            @app_path = [@app_path] unless @app_path.kind_of?(Array)
            @public_path = public_path || ($cuca_path + "/public")
            @log_path    = log_path || ($cuca_path + "/log")
            @app_path.freeze
            @public_path.freeze
            @log_path.freeze
            @logger      = Logger.new("#{@log_path}/messages")
            @logger.level = App.config['log_level'] || Logger::WARN
            @additional_support_directory = clean_path(additional_support_directory) if additional_support_directory
            @urlmap = Cuca::URLMap2.new do |config|
                config.base_path = @app_path
            end
        end

        # this will yield all support directories from base path and if defined
        # the naming proc else nil
        def all_support_directories(path_tree)
            path_tree.each do |t|
                (App::config['include_directories'] || []).each do |id|
                    if id.instance_of?(Hash) then
                        yield "#{t}/#{id[:dir]}", id[:class_naming]
                    else
        	            yield "#{t}/#{id}", nil
                    end
                end
            end
        end

        def load_support_files(scan_result)  # :nodoc:
            scan_result.path_tree.each do |pt|
                (App::config['include_directories'] || []).each do |id|
                    dir = id[:dir]
                    pro = id[:class_naming]
            
                    load_path = pt.children.find { |c| c.name == dir }      # raw finder
                    if load_path then
                        load_path.children.each do |file|
                            next unless file.value[:type] == :file
                            if pro then 
                                autoload_file(file, pro)
                            else
                                raise "not implemented"
                                include_files(file.value[:path])
                            end
                        end
                    end
                end
            end
        end
    
        # this will build an error message depending on configuration
        def get_error(title, exception, show_trace = true, file=nil)
            err = "<h3>#{title}</h3>"
            err << "<b>#{exception.class.to_s}: #{CGI::escapeHTML(exception.to_s)}</b><br/><br/>"
#           $stderr.puts "ERROR: #{title} - #{exception.class.to_s}: #{exception.to_s}"
            if (show_trace) then
                exception.backtrace.each do |b|
#                   $stderr.puts "   #{b}"
                    err +="<br/>#{b}"
                end
            else
                begin
                    err = File.open(file).read
                rescue
                end
            end
            err
        end


        def status2code(cgi_status)
            x = CGI::HTTP_STATUS[cgi_status]
            x||= "500 unknown status: #{cgi_status.inspect}"
            x.split(' ').first.to_i
        end

        def sanitize_headers(headers)    
            if headers['cookie'] then 
                headers['set-cookie'] = headers['cookie'].to_s
                headers.delete('cookie')
            end
            if headers['type'] then 
                headers['content-type'] = headers['type'].to_s
                headers.delete('type')
            end
      
            headers
        end

        def rack_response(code, headers, content)
            [code, sanitize_headers(headers), [content]]
        end

        def rackcall(env, after_init:nil)
            $cuca.value = {}
            $cuca.value[:app] = self
            $cuca.value[:request] = Rack::Request.new(env)
            $cuca.value[:env] = env

            #
            # 1st priority: Serve a file if it exists in the 'public' folder
            #
            file = @public_path + '/' + $cuca.value[:request].path_info
            if File.exists?(file) && File.ftype(file) == 'file' then
                require 'cuca/mimetypes'
                mt = MimeTypes.new
                file_content = File.open(file) { |f| f.read }
                extension = file.scan(/.*\.(.*)$/)[0][0] if file.include?('.')
                extension ||= 'html'
                mime = mt[extension] || 'text/html'
                return rack_response(200, 
                    { 'type' => mime, 
                    'expires' => (Time.now+App.config['http_static_content_expires']).to_s }, 
                    file_content)
            end
 
            begin 
                url_scan = @urlmap.scan($cuca.value[:request].path_info)
                $cuca.value[:url_scan] = url_scan
            rescue RoutingError => e
                err = get_error("Routing Error", e,
                Cuca::App.config['display_errors'], Cuca::App.config['http_500'])
                logger.info "CGICall Syntax Error"
                return rack_response(404, {'content-type' => 'text/html'}, err)
            end
 
            logger.info "RackCall on #{url_scan.url} - #{url_scan.script}"
            script = $cuca.value[:url_scan].script
 
            #
            # 2nd: Check if we have a script for requested action
            #
            if (!File.exists?(script)) then
              return [500, {}, ["Script not found: #{script}"]]
            end
        
            # 3rd: Load additional files
            load_support_files($cuca.value[:url_scan])
 
            # allow external code to be executed at this point
            if after_init then 
                after_init.call
            end
        
            # 4th: Now let's run the actual page script code
            if Cuca::App.config['controller_naming'] then
                controller_class_name = Cuca::App.config['controller_naming'].call($cuca.value[:url_scan].action)
            else
                controller_class_name = url_scan.action.capitalize+"Controller"
            end
 
            # Clear all hints
            Widget::clear_hints()

            # Load the code of the action into the module
            controller_module = $cuca.value[:url_scan].action_module
 
 
            # things fail in this block get error logged and/or displayed in browser
            begin
                # load controller
                begin
                    controller_module.module_eval(File.read(script), script)  unless \
                                controller_module.const_defined?(controller_class_name.intern)
                rescue SyntaxError,LoadError => e
                    err = get_error("Can not load script", e,
                    Cuca::App.config['display_errors'], Cuca::App.config['http_500'])
                    return [500, {}, [err]]
                end
           
                # Catch a common user error
                raise Cuca::ApplicationException.new("Could not find #{controller_class_name} defined in #{script}") \
                    unless controller_module.const_defined?(controller_class_name.intern)
                     
                    
                # run controller
                (status, mime, content, headers) = Sandbox.run(controller_class_name,
                                    controller_module, $cuca.value[:url_scan].assigns,
                                    $cuca.value[:request].request_method, $cuca.value[:url_scan].subcall)

                logger.debug "RackCall OK: #{status}/#{mime}"

                #raise  headers.merge( { 'type' => mime, 'status' => status}).inspect
                headers =  headers.merge( { 'content-type' => mime})
                code = self.status2code(status)
                return rack_response(code, headers, content)

            rescue SyntaxError => e
                err = get_error("Syntax Error", e,
                        Cuca::App.config['display_errors'], Cuca::App.config['http_500'])
                logger.info "CGICall Syntax Error"
                return rack_response(500, {}, err)
 
 
            rescue Cuca::ApplicationException => e
                err = get_error("Application Error", e,
                        Cuca::App.config['display_errors'], Cuca::App.config['http_500'])
 
                logger.info "CGICall Application Error"
                return rack_response(500, {'content-type' => 'text/html'}, err)
 
            rescue => e
                err = get_error("System Error", e,
                        Cuca::App.config['display_errors'], Cuca::App.config['http_500'])
                logger.info "CGICall System Error"
                return rack_response(500, {'content-type' => 'text/html'}, err)
            
            end #rescue block
        end # rackcall
    end # App
end # module
